src/Columns/IColumn.cpp

```diff
@@ -94,4 +109,363 @@ bool isColumnConst(const IColumn & column)
+    size_t num_sampled_rows = std::min(static_cast<size_t>(num_rows * sample_ratio), num_rows);
+    size_t num_checked_rows = 0;
+    size_t res = 0;
+
+    if (num_sampled_rows == num_rows || num_rows <= max_number_of_rows_for_full_search)
+    {
+        for (size_t i = 0; i < num_rows; ++i)
+            res += self.isDefaultAt(i);
+        num_checked_rows = num_rows;
+    }
+    else if (num_sampled_rows != 0)
+    {
+        for (size_t i = 0; i < num_rows; ++i)
+        {
+            if (num_checked_rows * num_rows <= i * num_sampled_rows)
+            {
+                res += self.isDefaultAt(i);
+                ++num_checked_rows;
+            }
+        }
+    }
+
+    if (num_checked_rows == 0)
+        return 0.0;
+
+    return static_cast<double>(res) / num_checked_rows;
+}
+
+template <typename Derived, typename Parent>
+UInt64 IColumnHelper<Derived, Parent>::getNumberOfDefaultRows() const
+{
+    const auto & self = static_cast<const Derived &>(*this);
+    UInt64 res = 0;
+    size_t num_rows = self.size();
+    for (size_t i = 0; i < num_rows; ++i)
+        res += self.isDefaultAt(i);
+    return res;
+}
+
+template <typename Derived, typename Parent>
+void IColumnHelper<Derived, Parent>::getIndicesOfNonDefaultRows(IColumn::Offsets & indices, size_t from, size_t limit) const
+{
+    const auto & self = static_cast<const Derived &>(*this);
+    size_t to = limit && from + limit < self.size() ? from + limit : self.size();
+    indices.reserve(indices.size() + to - from);
+
+    for (size_t i = from; i < to; ++i)
+    {
+        if (!self.isDefaultAt(i))
+            indices.push_back(i);
+    }
+}
+
+template <typename Derived, typename Parent>
+StringRef
+IColumnHelper<Derived, Parent>::serializeValueIntoArenaWithNull(size_t n, Arena & arena, char const *& begin, const UInt8 * is_null) const
+{
+    const auto & self = static_cast<const Derived &>(*this);
+    if (is_null)
+    {
+        char * memory;
+        if (is_null[n])
+        {
+            memory = arena.allocContinue(1, begin);
+            *memory = 1;
+            return {memory, 1};
+        }
+
+        size_t sz = self.byteSizeAt(n) + 1 /* null byte */;
+        memory = arena.allocContinue(sz, begin);
+        StringRef ret(memory, sz);
+        *memory = 0;
+        ++memory;
+        self.serializeValueIntoMemory(n, memory);
+        return ret;
+    }
+    else
+    {
+        return self.serializeValueIntoArena(n, arena, begin);
+    }
+}
+
+template <typename Derived, typename Parent>
+StringRef IColumnHelper<Derived, Parent>::serializeValueIntoArena(size_t n, Arena & arena, char const *& begin) const
+{
+    if constexpr (!std::is_base_of_v<ColumnFixedSizeHelper, Derived>)
+        return IColumn::serializeValueIntoArena(n, arena, begin);
+
+    const auto & self = static_cast<const Derived &>(*this);
+    size_t sz = self.byteSizeAt(n);
+    char * memory = arena.allocContinue(sz, begin);
+    self.serializeValueIntoMemory(n, memory);
```

Comment:
maybe I'm a little bit paranoid, but let's pls check that it is actually `ptr_returned_from_alloc + sz`

Reply:
`serializeValueIntoMemory` is changed to accept pointer value and return pointer to the end of the serialization data. It should be less prone to errors.